// graphql-anywhere/lib/src/fragments.js
"use strict";
var index_1 = require('./index');
function filter(fragment, data) {
    var resolver = function (fieldName, root, args, context, info) {
        return root[info.resultKey];
    };
    return index_1.default(resolver, fragment, data);
}
exports.filter = filter;
function check(fragment, data) {
    var resolver = function (fieldName, root, args, context, info) {
        if (!{}.hasOwnProperty.call(root, info.resultKey)) {
            throw new Error(info.resultKey + " missing on " + root);
        }
        return root[info.resultKey];
    };
    index_1.default(resolver, fragment, data, {}, {}, {
        fragmentMatcher: function () { return false; },
    });
}
exports.check = check;
var ANONYMOUS = '<<anonymous>>';
function PropTypeError(message) {
    this.message = message;
    this.stack = '';
}
PropTypeError.prototype = Error.prototype;
var reactPropTypeLocationNames = {
    prop: 'prop',
    context: 'context',
    childContext: 'child context',
};
function createChainableTypeChecker(validate) {
    function checkType(isRequired, props, propName, componentName, location, propFullName) {
        componentName = componentName || ANONYMOUS;
        propFullName = propFullName || propName;
        if (props[propName] == null) {
            var locationName = reactPropTypeLocationNames[location];
            if (isRequired) {
                if (props[propName] === null) {
                    return new PropTypeError(("The " + locationName + " `" + propFullName + "` is marked as required ") +
                        ("in `" + componentName + "`, but its value is `null`."));
                }
                return new PropTypeError(("The " + locationName + " `" + propFullName + "` is marked as required in ") +
                    ("`" + componentName + "`, but its value is `undefined`."));
            }
            return null;
        }
        else {
            return validate(props, propName, componentName, location, propFullName);
        }
    }
    var chainedCheckType = checkType.bind(null, false);
    chainedCheckType.isRequired = checkType.bind(null, true);
    return chainedCheckType;
}
function propType(fragment) {
    return createChainableTypeChecker(function (props, propName) {
        var prop = props[propName];
        try {
            check(fragment, prop);
            return null;
        }
        catch (e) {
            return e;
        }
    });
}
exports.propType = propType;
